apps/src/blockly/addons/cdoFieldButton.js

```diff
@@ -1,116 +1,77 @@
 import GoogleBlockly from 'blockly/core';
 
-const CORNER_RADIUS = 3;
-const INNER_HEIGHT = 16;
-
+/**
+ * This is a customized field which the user clicks to select an option from a customized picker,
+ * for example, the location of a sprite from a grid or a sound file from a customized modal.
+ * @param value The initial value of the field.
+ * @param validator A function that is called to validate changes to the field's value.
+ * Takes in a value & returns a validated value, or null to abort a change
+ * @param onChange The function that handles the field's editor.
+ * @param onDisplay The function tht handles how the field text is displayed.
+ * @param icon SVG <tspan> element - if the field displays a button, this is the icon that is displayed on the button.
+ */
 export default class CdoFieldButton extends GoogleBlockly.Field {
-  // Third argument (color) is used by CDO blockly, so we include it in
-  // method signature for compatibility
-  constructor(title, opt_buttonHandler, _, opt_changeHandler) {
-    super('');
-
-    this.title_ = title;
-    this.buttonHandler_ = opt_buttonHandler;
-    this.changeHandler_ = opt_changeHandler;
+  constructor(value, validator, onChange, onDisplay, icon) {
+    super(value, validator);
+    this.onChange = onChange;
+    this.onDisplay = onDisplay;
+    this.icon = icon;
+    this.SERIALIZABLE = true;
   }
 
-  /**
-   * @override
-   */
-  init() {
-    super.init();
-
-    this.buttonElement_ = Blockly.utils.dom.createSvgElement(
-      'rect',
-      {
-        rx: CORNER_RADIUS,
-        ry: CORNER_RADIUS,
-        x: 1,
-        y: 1,
-        height: INNER_HEIGHT,
-        width: INNER_HEIGHT,
-      },
-      this.fieldGroup_
+  static fromJson(options) {
+    return new CdoFieldButton(
+      options.value,
+      options.validator,
+      options.onChange,
+      options.onDisplay,
+      options.icon
     );
-    this.buttonElement_.style.fillOpacity = 1;
-    this.buttonElement_.style.fill =
-      this.getSourceBlock().style.colourSecondary;
-
-    this.textElement_.style.fontSize = '11pt';
-    this.textElement_.style.fill = this.getSourceBlock().style.colourPrimary;
-    this.textElement_.textContent = '';
-    this.textElement_.appendChild(this.title_);
-
-    this.fieldGroup_.insertBefore(this.buttonElement_, this.textElement_);
   }
 
   /**
-   * Used to update the value of a field.
+   * Create the block UI for this field.
    * @override
-   * @param newValue The value to be saved.
    */
-  doValueUpdate_(newValue) {
-    if (this.value_ !== newValue) {
-      if (this.changeHandler_) {
-        const override = this.changeHandler_(newValue);
-        if (override !== undefined) {
-          newValue = override;
-        }
-      }
-      this.value_ = newValue;
-      this.isDirty_ = true; // Block needs to be re-rendered.
+  initView() {
+    super.initView();
+    if (this.icon) {
+      this.icon.textContent = ' ' + this.icon.textContent;
```

Comment:
what is the extra space for?

Reply:
To add a space between the value and the icon. If I remove space, it looks like the following for the location picker:
<img width="328" alt="Screen Shot 2023-05-04 at 9 50 24 AM" src="https://user-images.githubusercontent.com/107423305/236245189-02ca260c-c111-49d5-9650-6a79416f51ad.png">

Reply:
Ah gotcha. I think it might be a bit clearer if the specific consumer of the field (location picker/sound picker) added the space in their text transformation code? By adding a space here, there's a bit of an assumption baked in that there will be something to the left of the icon, when we may in the future support a button field that only has an icon.

Reply:
Good point! I like the idea that this class can support a button that only has an icon. Easy to add space in `blocks.js`. Thanks!